--- /usr/share/emacs/site-lisp/magit.orig.el	2010-03-23 11:58:05.000000000 +0900
+++ /usr/share/emacs/site-lisp/magit.el	2011-08-22 16:12:43.515462736 +0900
@@ -987,10 +987,19 @@
     (magit-set-mode-line-process nil)
     (magit-refresh-buffer magit-process-client-buffer)))
 
+(defun magit-password (proc string)
+  "Checks if git/ssh asks for a password and ask the user for it."
+  (when (string-match "^Enter passphrase for key '\\\(.*\\\)': $" string)
+    (process-send-string proc
+                         (concat (read-passwd
+                                  (format "Password for '%s':" (match-string 1 string))
+                                  nil) "\n"))))
+
 (defun magit-process-filter (proc string)
   (save-current-buffer
     (set-buffer (process-buffer proc))
     (let ((inhibit-read-only t))
+      (magit-password proc string)
       (goto-char (process-mark proc))
       ;; Find last ^M in string.  If one was found, ignore everything
       ;; before it and delete the current line.
